src/calliope/model.py

```diff
@@ -223,21 +220,25 @@ def _init_from_model_def_dict(
             init_config, model_definition, data_sources, attributes, param_metadata
         )
         model_data_factory.build()
-
         self._model_data = model_data_factory.dataset
-        self._model_data.attrs["math"] = self._add_math(init_config["add_math"])
 
         log_time(
             LOGGER,
             self._timings,
             "model_data_creation",
             comment="Model: preprocessing stage 2 (model_data)",
         )
+
+        math, applied_math = self._math_init_from_yaml(init_config)
```

Comment:
The new `_math`  functions follow a functional programming approach for two reasons:
- easier to write tests for this style
- is explicit

I can change this to an implicit OOP approach if need-be.